# Pipeline to analyze rDNA variant calling pipeline performance
This GitHub repository contains all the scripts used to process data and generate my Final Degree Project results. 

> NOTE: no data is available in this repository. 

## Repository Structure
The folders in this repository are organized based on the order of the analysis. Inside each folder, you will find a README file that provides information on the procedure conducted, *IMAGES* and the *scripts* used.

### Folders
#### 1. Kmers
- **Description**: Analysis of the performance of the pipeline using reads without errors.
#### 2. Simulated reads
- **Description**: Analysis of the preprocessing step of the pipeline using simulated reads with sequencing errors.
#### 3. VC simulated
- **Description**: Analysis of the performance of the variant calling pipeline testing different parameters.
#### 4. rDNAs
- **Description**: Identification isolation and alignment of all sets of unique rDNA sequences currently available.
#### 6. Validation
- **Description**: Variant calling of the best parameters of each version of the pipeline on the T2T-CHM13 WGS data. 
#### 8. Performance 
- **Description**: Quantification of the precision, recall, and F1 score of all tested parameters.

## Contact
    Email: alba.jimenez@alum.esci.upf.edu
    GitHub: https://github.com/AlbaJimLup
